File: css/money/ch15/Money.Tests/MoneyTest.cs
using NUnit.Framework;
using TheSoftwareGorilla.TDD.Money;

namespace TheSoftwareGorilla.TDD.Money.Tests;

[TestFixture]
public class MoneyTests
{
    private Money? _fiveDollar;

    [SetUp]
    public void Setup()
    {
        _fiveDollar = Money.Dollar(5);
    }

    [TearDown]
    public void TearDown()
    {
        _fiveDollar = null;
    }


    [Test]
    public void TestConstruction()
    {
        Assert.IsNotNull(_fiveDollar);
        Assert.That(_fiveDollar.Amount, Is.EqualTo(5));
        Assert.That(_fiveDollar.Currency, Is.EqualTo("USD"));

        var fiveFranc = Money.Franc(5);
        Assert.IsNotNull(fiveFranc);
        Assert.That(fiveFranc.Amount, Is.EqualTo(5));
        Assert.That(fiveFranc.Currency, Is.EqualTo("CHF"));

        // Had to add this test because I'm South African! :)
        var fiveRand = Money.Rand(5);
        Assert.IsNotNull(fiveRand);
        Assert.That(fiveRand.Amount, Is.EqualTo(5));
        Assert.That(fiveRand.Currency, Is.EqualTo("ZAR"));
    }


    [Test]
    public void TestEquality()
    {
        Assert.That(_fiveDollar, Is.EqualTo(Money.Dollar(5)));
        Assert.That(Money.Dollar(6), Is.Not.EqualTo(Money.Dollar(5)));
        Assert.That(Money.Franc(5), Is.Not.EqualTo(Money.Dollar(5)));
   }


   [Test]
   public void TestCurrency()
   {
        Assert.That(Money.Dollar(1).Currency, Is.EqualTo("USD"));
        Assert.That(Money.Franc(1).Currency, Is.EqualTo("CHF"));
        Assert.That(Money.Rand(1).Currency, Is.EqualTo("ZAR"));
   }

    [Test]
    public void TestMultiplication()
    {
        Assert.That(Money.Dollar(10), Is.EqualTo(_fiveDollar.Times(2)));
        Assert.That(Money.Dollar(15), Is.EqualTo(_fiveDollar.Times(3)));
    }

    [Test]
    public void TestSimpleAddition()
    {
        Expression sum = _fiveDollar.Plus(_fiveDollar);
        Bank bank = new Bank();
        Money reduced = bank.Reduce(sum, "USD");
        Assert.That(Money.Dollar(10), Is.EqualTo(reduced));
    }

    [Test]
    public void TestPlusReturnsSum()
    {
        Expression result = _fiveDollar.Plus(_fiveDollar);
        Sum sum = (Sum) result;
        Assert.That(_fiveDollar, Is.EqualTo(sum.Augend));
        Assert.That(_fiveDollar, Is.EqualTo(sum.Addend));
    }

    [Test]
    public void TestReduceSum()
    {
        Expression sum = new Sum(Money.Dollar(3), Money.Dollar(4));
        Bank bank = new Bank();
        Money result = bank.Reduce(sum, "USD");
        Assert.That(Money.Dollar(7), Is.EqualTo(result));
    }

    [Test]
    public void TestReduceMoney()
    {
        Bank bank = new Bank();
        Money result = bank.Reduce(Money.Dollar(1), "USD");
        Assert.That(Money.Dollar(1), Is.EqualTo(result));
    }

    [Test]
    public void TestReduceMoneyDifferentCurrency()
    {
        Bank bank = new Bank();
        bank.AddRate("CHF", "USD", 2);
        Money randExchange = bank.Reduce(Money.Franc(2), "USD");
        Assert.That(Money.Dollar(1), Is.EqualTo(randExchange));
        bank.AddRate("ZAR", "USD", 17);
        randExchange = bank.Reduce(Money.Rand(85), "USD");
        Assert.That(randExchange, Is.EqualTo(_fiveDollar));
    }

    // On page 69, Kent suggests trying this to see if the test will pass. In 2002, Java did not have 
    // a built-in array equality method test, so this code would not compile.
    // The Java code he suggested then is:
    // assertEquals(new Object[] {"abc"}, new Object[] {"abc"}); 
    // I included this same test in the Java example, but used assertArrayEquals instead.
    // NUnit supports this in 2024, so this test will pass and I am leaving it here for reference.
    [Test]
    public void TestArrayEquals()
    {
        var actual = new object[] { "abc" };
        Assert.That(actual, Is.EqualTo(new object[] { "abc" }));
    }

    // This code is not in the book, but when I started adding the CurrencyPair class, I realized we should 
    // have started with a test.
    [Test]
    public void TestCurrencyPairEquals()
    {
        var FrancToDollar = new CurrencyPair("CHF", "USD");
        Assert.That(FrancToDollar, Is.EqualTo(new CurrencyPair("CHF", "USD")));
        Assert.That(FrancToDollar.GetHashCode(), Is.EqualTo(new CurrencyPair("CHF", "USD").GetHashCode()));
    }

    [Test]
    public void TestIdentityRate()
    {
        Assert.That(new Bank().Rate("USD", "USD"), Is.EqualTo(1));
        Assert.That(new Bank().Rate("GBP", "ZAR"), Is.EqualTo(0)); // We haven't added this rate to the rate table so make sure it is zero.
    }

    [Test]
    public void TestMixedAddition()
    {
        Expression fiveBucks = _fiveDollar;
        Expression tenFrancs = Money.Franc(10);
        Bank bank = new Bank();
        bank.AddRate("CHF", "USD", 2);
        Money result = bank.Reduce(fiveBucks.Plus(tenFrancs), "USD");
        Assert.That(result, Is.EqualTo(Money.Dollar(10)));
    }


}

File: java/ch14/src/test/java/com/thesoftwaregorilla/tdd/money/MoneyTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class MoneyTest {

    private final Money fiveDollar = Money.dollar(5);

    // This code is not in the book, but my regular tests always have a construction test
    // In chapter 4, when we made amount private, we needed to add a getter in the Dollar class.
    // I could have removed this test because it is tested anyway in the multiplication test, but
    // I like to have a construction test.
    // Also, I wanted to keep parity with the C# code where the Amount member is read only.
    @Test
    public void testConstruction() {
        assertNotNull(fiveDollar);
        assertEquals(5, fiveDollar.getAmount());
        assertEquals("USD", fiveDollar.getCurrency());

        Money fiveFranc = Money.franc(5);
        assertNotNull(fiveFranc);
        assertEquals(5, fiveFranc.getAmount());
        assertEquals("CHF", fiveFranc.getCurrency());

        // Had to add this because I'm South African :)
        Money fiveRand = Money.rand(5);
        assertNotNull(fiveRand);
        assertEquals(5, fiveRand.getAmount());
        assertEquals("ZAR", fiveRand.getCurrency());

    }

    @Test
    public void testEquality() {
        assertTrue(fiveDollar.equals(Money.dollar(5)));
        assertFalse(fiveDollar.equals(Money.dollar(6)));
        assertFalse(Money.franc(5).equals(fiveDollar));
    }


    @Test
    public void testCurrency() {
        assertEquals("USD", Money.dollar(1).getCurrency());
        assertEquals("CHF", Money.franc(1).getCurrency());
        assertEquals("ZAR", Money.rand(1).getCurrency());
    }

    @Test
    public void testMultiplication() {
        assertEquals(Money.dollar(10), fiveDollar.times(2));
        assertEquals(Money.dollar(15), fiveDollar.times(3));
    }

    @Test
    public void testSimpleAddition() {
        Expression sum = fiveDollar.plus(fiveDollar);
        Bank bank = new Bank();
        Money reduced = bank.reduce(sum, "USD");
        assertEquals(Money.dollar(10), reduced);
    }

    @Test
    public void testPlusReturnsSum() {
        Expression result = fiveDollar.plus(fiveDollar);
        Sum sum = (Sum) result;
        assertEquals(fiveDollar, sum.augend);
        assertEquals(fiveDollar, sum.addend);
    }

    @Test
    public void testReduceSum() {
        Expression sum = new Sum(Money.dollar(3), Money.dollar(4));
        Bank bank = new Bank();
        Money result = bank.reduce(sum, "USD");
        assertEquals(Money.dollar(7), result);
    }

    @Test
    public void testReduceMoney() {
        Bank bank = new Bank();
        Money result = bank.reduce(Money.dollar(1), "USD");
        assertEquals(Money.dollar(1), result);
    }

    @Test
    public void testReduceMoneyDifferentCurrency() {
        Bank bank = new Bank();
        bank.addRate("CHF", "USD", 2); // 1 CHF = 0.5 USD
        Money result = bank.reduce(Money.franc(2), "USD");
        assertEquals(Money.dollar(1), result);
        bank.addRate("ZAR", "USD", 17); // 17 ZAR = 1 USD
        Money randExchange = bank.reduce(Money.rand(85), "USD");
        assertEquals(fiveDollar, randExchange);
    }

    // On page 69, Kent suggests trying this to see if the test will pass. In 2002, Java did not have
    // a built-in array equality method test, so the code in the book would not compile, and still fails as a test.
    // The Java code he suggested then is:
    // assertEquals(new Object[] {"abc"}, new Object[] {"abc"});
    // JUnit 5 supports array comparison in 2024, so this test will pass and I am leaving it here for reference.
    @Test
    public void testArrayEquals() {
//        assertEquals(new Object[] {"abc"}, new Object[] {"abc"}); // This test will fail.
        var expected = new Object[] { "abc" };
        var actual = new Object[] { "abc" };
        assertArrayEquals(actual, expected);
    }

    // This code is not in the book, but when I started adding the CurrencyPair class, I realized we should have
    // started with a test.
    @Test
    public void testCurrencyPairEquals() {
        var francToDollar = new CurrencyPair("CHF", "USD");
        assertEquals(francToDollar, new CurrencyPair("CHF", "USD"));
        assertEquals(francToDollar.hashCode(), new CurrencyPair("CHF", "USD").hashCode());
    }

    @Test
    public void testIdentityRate() {
        assertEquals(1, new Bank().rate("USD", "USD"));
        assertEquals(0, new Bank().rate("GBP", "ZAR")); // No rate defined, should return 0
    }

}
